pkg/controller/clusterserviceversion/clusterserviceversion_controller.go

```diff
@@ -196,5 +278,98 @@ func (r *ReconcileClusterServiceVersion) Reconcile(request reconcile.Request) (r
+	err = r.client.List(context.TODO(), meterDefinitionList, client.InNamespace(CSV.GetNamespace()))
+	if err != nil {
+		return err
+	}
+
+	// search specifically for the Meter Definition created by the CSV
+	for _, meterDefinition := range meterDefinitionList.Items {
+		ann := meterDefinition.GetAnnotations()
+		if _, ok := ann[utils.CSV_ANNOTATION_NAME]; ok {
+			// CSV specific MeterDefinition found; delete it
+			err = r.client.Delete(context.TODO(), &meterDefinition, client.PropagationPolicy(metav1.DeletePropagationForeground))
+			if err != nil {
+				return err
+			}
+			reqLogger.Info("found and deleted MeterDefinition")
+			return nil
+		}
+	}
+
+	reqLogger.Info("no meterdefinition found")
+	return nil
+}
+
+// checkAnnotations checks the Annotations for the rhm CSV
+// If the CSV is new, we tag it and create a MeterDefinition
+// If the CSV is old, we check if the existing MeterDefinition matches the CSV (json) MeterDefinition
+func (r *ReconcileClusterServiceVersion) checkAnnotations(CSV *olmv1alpha1.ClusterServiceVersion, annotations map[string]string) error {
+	var err error
+	reqLogger := log.WithValues("CSV.Name", CSV.Name, "CSV.Namespace", CSV.Namespace)
+
+	// retrives the string representatin of the MeterDefinition from annotations
+	// and builds a MeterDefinition instance out of it
+	reqLogger.Info("retrieving MeterDefinition string from csv")
+	meterDefinitionString, ok := annotations[utils.CSV_METERDEFINITION_ANNOTATION]
+	if ok {
+		reqLogger.Info("retrieval successful")
+	} else {
+		err = goerr.New("could not retrieve MeterDefinitionString")
+		reqLogger.Error(err, "was expecting a an annotation for meterdefinition in CSV")
+	}
+	reqLogger.Info("building a local copy of MeterDefinition")
+	meterDefinition := &marketplacev1alpha1.MeterDefinition{}
+	_, err = meterDefinition.BuildMeterDefinitionFromString(meterDefinitionString, CSV.GetName(), CSV.GetNamespace(), utils.CSV_ANNOTATION_NAME, utils.CSV_ANNOTATION_NAMESPACE)
+	if err != nil {
+		reqLogger.Error(err, "Could not build a local copy of the MeterDefinition")
+		return err
+	}
+	reqLogger.Info("Checking if the csv is being tracked")
+	//checks if the CSV is new or old (aka. is it already being tracked)
+	if val, ok := annotations[trackMeterTag]; ok && val == "true" {
+		// Case 1: The CSV is old: no action required
+		reqLogger.Info("CSV is already tracked")
+		existingMeterDefinition := &marketplacev1alpha1.MeterDefinition{}
+		err = r.client.Get(context.TODO(), client.ObjectKey{Name: meterDefinition.GetName(), Namespace: meterDefinition.GetNamespace()}, existingMeterDefinition)
+		if err != nil {
+			reqLogger.Error(err, "Could not retrieve the existing MeterDefinition")
+			return err
+		}
+		if !reflect.DeepEqual(meterDefinition.Spec, existingMeterDefinition.Spec) && !reflect.DeepEqual(meterDefinition.ObjectMeta, existingMeterDefinition.ObjectMeta) {
+			err = goerr.New("Existing MeterDefinition and Expected MeterDefinition mismatch")
```

Comment:
Suggestion (no action required): Use emperror/errors instead of goerrs. It'll include line numbers and other stack trace info for you.